### PR TITLE
optimization 3

### PR DESCRIPTION
**Pull request by gandrewstone**
Builds on #1485 
give mapBlockIndex its own lock rather than using cs_main & related cleanup

### CODE REVIEW

File: src/txadmission.cpp [hunk 1]
```diff
@@ -442,7 +443,7 @@ void ThreadTxAdmission()
                             static const unsigned int nMaxOrphanTx = (unsigned int)std::max(
                                 (int64_t)0, GetArg("-maxorphantx", DEFAULT_MAX_ORPHAN_TRANSACTIONS));
                             static const uint64_t nMaxOrphanPoolSize = (uint64_t)std::max(
-                                (int64_t)0, (GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000 / 10));
+                                (int64_t)0, (GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000 / 2));
                             unsigned int nEvicted = orphanpool.LimitOrphanTxSize(nMaxOrphanTx, nMaxOrphanPoolSize);
```
Comment (ptschip) [inline src/txadmission.cpp]: not sure i agree with making the orphanpool so big.  I feel like we could open up an out of memory attack vector here, or possibly cause nodes with less available memory to run out of memory.  If we want to change this then I think we should have a separate PR which gives us a new option -maxorphanpool, so the users can override and configure it if necessary.

File: src/blockrelay/thinblock.cpp [hunk 2]
```diff
@@ -484,40 +478,37 @@ bool CXRequestThinBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
...
-        if (mi == mapBlockIndex.end())
+        dosMan.Misbehaving(pfrom, 20);
+        return error("Requested block is not available");
+    }
+    else
+    {
+        CBlock block;
+        const Consensus::Params &consensusParams = Params().GetConsensus();
+        if (!ReadBlockFromDisk(block, hdr, consensusParams))
         {
```
Comment (ptschip) [inline src/blockrelay/thinblock.cpp:493]: I believe we still need cs_main for reading/writing blocks to disk.  Level db is not really thread safe from my experience with the utxo.  This is another area we could apply a different criticalsection for doing these reads/writes.

File: src/main.cpp [hunk 3]
```diff
@@ -2025,17 +2033,16 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
...
             {
                 dosMan.Misbehaving(pfrom, 100);
                 return error("Peer %srequested nonexistent block %s", pfrom->GetLogName(), inv.hash.ToString());
             }
 
             CBlock block;
             const Consensus::Params &consensusParams = Params().GetConsensus();
-            if (!ReadBlockFromDisk(block, (*mi).second, consensusParams))
+            if (!ReadBlockFromDisk(block, invIndex, consensusParams))
             {
```
Comment (ptschip) [inline src/main.cpp:2046]: same here, i think we need cs_main or some other lock for ReadBlockFromDisk

File: src/txadmission.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/txadmission.cpp]: this is sized for giga-blocks.  you are right we are far from that on the real network.  I'll revert.

File: src/blockrelay/thinblock.cpp (hunk 2 above)
Comment (gandrewstone) [reply src/blockrelay/thinblock.cpp:493]: see https://github.com/google/leveldb/blob/master/doc/index.md#concurrency.
Something's not thread safe in the utxo then... maybe bitcoind's cache?

But also most people are still using the "flat" block access which should be thread safe because written files are immutable.  So if blockdb turns out to be not thread safe or leveldb buggy we should lock it at a lower level.  

(also, I've been running 2 nodes one on each fork for a few days now with no issues)

**Comment by ptschip** [comment]
the utxo is always locked with cs_utxo...that's something we had to do.
but i agree , most people are using flat files right now so shouldn't be
an issue.  I can open a PR when you're away , simple enough to add
somelocking in blockstorage.cpp where we read and write data there.

**Comment by ptschip** [comment]
Looks good but you'll need to drop the first couple of commits, once optimization2 is merged.

**Comment by sickpig** [comment]
`checkdatasig` and `invalidtxrequest` python tests are failing with the same assertion: "Not all nodes requested block" (see https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/458243364)

edit: I'm not able to reproduce these failures locally

**Comment by sickpig** [comment]
The failures happen in `qa/rpc-tests/test_framework/comptool.py`

```py
# --> error if not requested
if not wait_until(blocks_requested, attempts=20*num_blocks):
    # print [ c.cb.block_request_map for c in self.connections ]
    raise AssertionError("Not all nodes requested block")
```

basically according to `wait_until` code we check for the return value of `blocks_requested()` being true 20 times before failing.

This means that not all nodes received all the mined blocks after a `sync_blocks()` call, seems like there's some kind of network problem in the travis instance that make slow for all blocks to propagate. 

Another rather strange thing I noticed is the way the error is reported to `stdout`:

```
stderr:
   File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py", line 222, in main
    self.run_test()
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/checkdatasig_activation.py", line 115, in run_test
    self.test.run()
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/comptool.py", line 370, in run
    self.sync_blocks(block.sha256, 1)
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/comptool.py", line 211, in sync_blocks
    raise AssertionError("Not all nodes requested block")
```

Basically the only reference to the file that contains the test is L115 which correspond to: 

```
self.test.run()
```

I would have expected something more specific or is it  `self.test.run()` the last line executed before the failure?

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -650,23 +647,23 @@ UniValue getblock(const UniValue &params, bool fHelp)
...
 
-    if (mapBlockIndex.count(hash) == 0)
+    CBlockIndex *pblockindex = LookupBlockIndex(hash);
+    if (!pblockindex)
     {
         arith_uint256 h = UintToArith256(hash);
         if (h.bits() < 65)
         {
+            LOCK(cs_main);
             uint64_t height = boost::lexical_cast<unsigned int>(strHash);
```
Comment (ptschip) [inline src/rpc/blockchain.cpp:667]: I don't think chainActive.Height() needs cs_main anymore...it relies on tip() which is atomic.

File: src/init.cpp [hunk 5]
```diff
@@ -1474,7 +1477,10 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
...
 
     //// debug print
-    LOGA("mapBlockIndex.size() = %u\n", mapBlockIndex.size());
+    {
+        READLOCK(cs_mapBlockIndex);
+        LOGA("mapBlockIndex.size() = %u\n", mapBlockIndex.size());
+    }
     {
         LOCK(cs_main);
         LOGA("nBestHeight = %d\n", chainActive.Height());
```
Comment (ptschip) [inline src/init.cpp]: another chainactive.height()

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (gandrewstone) [reply src/rpc/blockchain.cpp:667]: its needed for "pblockindex = chainActive[height];" just below.  

giving chainActive its own lock is the next thing to do for multithread perf I think...

File: src/init.cpp (hunk 5 above)
Comment (gandrewstone) [reply src/init.cpp]: thx, fixed

**Review by ptschip — APPROVED** [review]
